File: tests/01-pymod_basics/005-add_1_to_arg/test_add_1_to_arg.nim
import pymod
import pymodpkg/pyarrayobject
import unicode


proc cfloatAdd1ToArg*(arg: cfloat):   cfloat {.exportpy.} = arg + 1.0  # No `succ`
proc cdoubleAdd1ToArg*(arg: cdouble): cdouble {.exportpy.} = arg + 1.0  # No `succ`

proc cshortAdd1ToArg*(arg: cshort):   cshort {.exportpy.} = succ(arg)
proc cintAdd1ToArg*(arg: cint):       cint {.exportpy.} = succ(arg)
proc clongAdd1ToArg*(arg: clong):     clong {.exportpy.} = succ(arg)

proc cushortAdd1ToArg*(arg: cushort): cushort {.exportpy.} = succ(arg)
proc cuintAdd1ToArg*(arg: cuint):     cuint {.exportpy.} = succ(arg)
proc culongAdd1ToArg*(arg: culong):   culong {.exportpy.} = arg + 1  # No `succ`

proc floatAdd1ToArg*(arg: float):     float {.exportpy.} = arg + 1.0  # No `succ`
proc float32Add1ToArg*(arg: float32): float32 {.exportpy.} = arg + 1.0  # No `succ`
proc float64Add1ToArg*(arg: float64): float64 {.exportpy.} = arg + 1.0  # No `succ`

proc intAdd1ToArg*(arg: int):         int {.exportpy.} = succ(arg)
#proc int8Add1ToArg*(arg: int8):       int8 {.exportpy.} = succ(arg)
proc int16Add1ToArg*(arg: int16):     int16 {.exportpy.} = succ(arg)
proc int32Add1ToArg*(arg: int32):     int32 {.exportpy.} = succ(arg)
proc int64Add1ToArg*(arg: int64):     int64 {.exportpy.} = succ(arg)

proc uintAdd1ToArg*(arg: uint):       uint {.exportpy.} = arg + 1  # No `succ`
proc uint8Add1ToArg*(arg: uint8):     uint8 {.exportpy.} = succ(arg)
proc uint16Add1ToArg*(arg: uint16):   uint16 {.exportpy.} = succ(arg)
proc uint32Add1ToArg*(arg: uint32):   uint32 {.exportpy.} = succ(arg)
proc uint64Add1ToArg*(arg: uint64):   uint64 {.exportpy.} = arg + 1  # No `succ`

#proc boolAdd1ToArg*(arg: bool):       bool {.exportpy.} = succ(arg)
proc byteAdd1ToArg*(arg: byte):       byte {.exportpy.} = succ(arg)

proc ccharAdd1ToArg*(arg: cchar):    cchar {.exportpy.} = succ(arg)
proc charAdd1ToArg*(arg: char):      char {.exportpy.} = succ(arg)
proc stringAdd1ToArg*(arg: string):  string {.exportpy.} = arg & "def"

#proc unicodeRuneAdd1ToArg*(arg: Rune):  Rune {.exportpy.} = succ(arg)
#proc seqCharAdd1ToArg*(arg: seq[char]): seq[char] {.exportpy.} = succ(arg)
#proc seqRuneAdd1ToArg*(arg: seq[Rune]): seq[Rune] {.exportpy.} = succ(arg)


initPyModule("",
    cfloatAdd1ToArg, cdoubleAdd1ToArg,
    cshortAdd1ToArg, cintAdd1ToArg, clongAdd1ToArg,
    cushortAdd1ToArg, cuintAdd1ToArg, culongAdd1ToArg,
    floatAdd1ToArg, float32Add1ToArg, float64Add1ToArg,
    #int8Add1ToArg,
    intAdd1ToArg, int16Add1ToArg, int32Add1ToArg, int64Add1ToArg,
    uintAdd1ToArg, uint8Add1ToArg, uint16Add1ToArg, uint32Add1ToArg, uint64Add1ToArg,
    #boolAdd1ToArg,
    byteAdd1ToArg,
    ccharAdd1ToArg, charAdd1ToArg, stringAdd1ToArg,
    #unicodeRuneAdd1ToArg, seqCharAdd1ToArg, seqRuneAdd1ToArg,
    )

File: tests/00-pymod_basics/005-add_1_to_arg/test_add_1_to_arg.py
def test_0_compile_pymod_test_mod(pmgen_py_compile):
    pmgen_py_compile(__name__)


def test_cfloatAdd1ToArg(pymod_test_mod):
    arg = 1.0
    res = pymod_test_mod.cfloatAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_cdoubleAdd1ToArg(pymod_test_mod):
    arg = 1.0
    res = pymod_test_mod.cdoubleAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


def test_cshortAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.cshortAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_cintAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.cintAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_clongAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.clongAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


def test_cushortAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.cushortAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_cuintAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.cuintAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_culongAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.culongAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


def test_floatAdd1ToArg(pymod_test_mod):
    arg = 1.0
    res = pymod_test_mod.floatAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_float32Add1ToArg(pymod_test_mod):
    arg = 1.0
    res = pymod_test_mod.float32Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_float64Add1ToArg(pymod_test_mod):
    arg = 1.0
    res = pymod_test_mod.float64Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


def test_intAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.intAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

#def test_int8Add1ToArg(pymod_test_mod):
#    arg = 1
#    res = pymod_test_mod.int8Add1ToArg(arg)
#    assert res == (arg + 1)
#    assert type(res) == type(arg)

def test_int16Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.int16Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_int32Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.int32Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_int64Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.int64Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


def test_uintAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.uintAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_uint8Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.uint8Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_uint16Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.uint16Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_uint32Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.uint32Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)

def test_uint64Add1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.uint64Add1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


#def test_boolAdd1ToArg(pymod_test_mod):
#    arg = True
#    res = pymod_test_mod.boolAdd1ToArg(arg)
#    assert res == (arg + 1)
#    assert type(res) == type(arg)

def test_byteAdd1ToArg(pymod_test_mod):
    arg = 1
    res = pymod_test_mod.byteAdd1ToArg(arg)
    assert res == (arg + 1)
    assert type(res) == type(arg)


def test_ccharAdd1ToArg(pymod_test_mod, python_major_version):
    # Python 3 or above: bytes vs strings, yay!
    arg = b"a" if python_major_version >= 3 else "a"
    expectedRes = b"b" if python_major_version >= 3 else "b"
    res = pymod_test_mod.ccharAdd1ToArg(arg)
    assert res == expectedRes
    assert type(res) == type(expectedRes)
    assert type(res) == type(arg)

def test_charAdd1ToArg(pymod_test_mod, python_major_version):
    # Python 3 or above: bytes vs strings, yay!
    arg = b"a" if python_major_version >= 3 else "a"
    expectedRes = b"b" if python_major_version >= 3 else "b"
    res = pymod_test_mod.charAdd1ToArg(arg)
    assert res == expectedRes
    assert type(res) == type(expectedRes)
    assert type(res) == type(arg)

def test_stringAdd1ToArg(pymod_test_mod):
    arg = "abc"
    expectedRes = arg + "def"
    res = pymod_test_mod.stringAdd1ToArg(arg)
    assert res == expectedRes
    assert type(res) == type(expectedRes)
    assert type(res) == type(arg)


#def test_unicodeRuneAdd1ToArg(pymod_test_mod, python_major_version):
#    arg = "a" if python_major_version >= 3 else u"a"
#    expectedRes = "b" if python_major_version >= 3 else u"b"
#    res = pymod_test_mod.unicodeRuneAdd1ToArg(arg)
#    assert res == expectedRes
#    assert type(res) == type(expectedRes)
#    assert type(res) == type(arg)

#def test_seqCharAdd1ToArg(pymod_test_mod, python_major_version):
#    # Python 3 or above: bytes vs strings, yay!
#    arg = b"abc" if python_major_version >= 3 else "abc"
#    expectedRes = (arg + b"def") if python_major_version >= 3 else (arg + "def")
#    res = pymod_test_mod.seqCharAdd1ToArg(arg)
#    assert res == expectedRes
#    assert type(res) == type(expectedRes)
#    assert type(res) == type(arg)

#def test_seqRuneAdd1ToArg(pymod_test_mod, python_major_version):
#    arg = "abc" if python_major_version >= 3 else u"abc"
#    expectedRes = (arg + "def") if python_major_version >= 3 else (arg + u"def")
#    res = pymod_test_mod.seqRuneAdd1ToArg(arg)
#    assert res == expectedRes
#    assert type(res) == type(expectedRes)
#    assert type(res) == type(arg)

